pontoon/base/static/js/fluent_interface.js

```diff
@@ -6,21 +6,136 @@ var fluentSerializer = new FluentSyntax.FluentSerializer();
 var Pontoon = (function (my) {
 
   /*
-   * Render original string element: simple string value or a variant.
+   * Is ast element of type that can be presented as a simple string:
+   * - TextElement
+   * - Placeable with expression type CallExpression, ExternalArgument
+   *   or MessageReference
    */
-  function renderOriginalElement(value, elements) {
+  function isSimpleElement(element) {
+    if (element.type === 'TextElement') {
+      return true;
+    }
+
+    // Placeable
+    if (
+      element.expression &&
+      [
+        'CallExpression',
+        'ExternalArgument',
+        'MessageReference'
+      ].indexOf(element.expression.type) >= 0
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /*
+   * Is ast element representing a pluralized string?
+   *
+   * Keys of all variants of such elements are either
+   * CLDR plurals or numbers.
+   */
+  function isPluralElement(element) {
+    if (!(element.expression && element.expression.type === 'SelectExpression')) {
+      return false;
+    }
+
+    var CLDRplurals = ['zero', 'one', 'two', 'few', 'many', 'other'];
+
+    return element.expression.variants.every(function (item) {
+      return (
+        CLDRplurals.indexOf(item.key.name) !== -1 ||
+        item.key.type === 'NumberExpression'
+      );
+    });
+  }
+
+  /*
+   * Is ast of a message, supported in rich FTL editor?
+   *
+   * Message is supported if it's valid and all value elements
+   * and all attribute elements are:
+   * - simple elements or
+   * - select expressions
+   */
+  function isSupportedMessage(ast) {
+    function elementsSupported(elements) {
+      return elements.every(function(element) {
+        return (
+          isSimpleElement(element) ||
+          (element.expression && element.expression.type === 'SelectExpression')
+        );
+      });
+    }
```

Comment:
As far as I can see, there's no point in defining this function inside another function. By doing that, you'll force the compiler to recreate the `elementsSupported` function every time you invoke `isSupportedMessage`, so it would be better to put it outside (even if it's only ever used inside `isSupportedMessage`).